karabo_bridge/simulation.py

```diff
@@ -29,125 +29,183 @@
+    @property
+    def data_type(self):
+        return np.float32 if self.corrected else np.uint16
 
-    # detector random data
-    if corrected:
-        gain_data = np.zeros(array_shape, dtype=np.uint16)
-        domain = detector_info['source'].partition('/')[0]
+    def corr_passport(self):
+        domain = self.source.partition('/')[0]
         passport = [
             '%s/CAL/THRESHOLDING_Q1M1' % domain,
             '%s/CAL/OFFSET_CORR_Q1M1' % domain,
             '%s/CAL/RELGAIN_CORR_Q1M1' % domain
         ]
+        return passport
+
+    def random(self):
+        return np.random.uniform(low=1500, high=1600,
+                                 size=self.data_shape).astype(self.data_type)
+
+    def zeros(self):
+        return np.zeros(self.data_shape, dtype=self.data_type)
+
+    def module_position(self, ix):
+        y, x = np.where(self.layout == ix)
+        assert len(y) == len(x) == 1
+        return x[0], y[0]
+
+    @staticmethod
+    def gen_metadata(source, timestamp, trainId):
+        sec, frac = str(timestamp).split('.')
+        meta = {}
+        meta[source] = {
+            'source': source,
+            'timestamp': timestamp,
+            'timestamp.tid': trainId,
+            'timestamp.sec': sec,
+            'timestamp.frac': frac.ljust(18, '0')  # attosecond resolution
+        }
+        return meta
+
+    def gen_data(self, trainId):
+        data = dict(self.const_fields)
+
+        timestamp = time()
+        img = self.genfunc()
+        base_src = '/'.join((self.source.rpartition('/')[0], '{}CH0:xtdf'))
+        sources = [base_src.format(i) for i in range(16)]
+
+        data['__intime'] = timestamp
+        # TODO: cellId differ between AGIPD/LPD
+        data['cellId'] = np.arange(self.pulses, dtype=np.uint16)
+        data['combinedFrom'] = sources
+        data['image.data'] = img
+        data['length'] = np.full((self.pulses, 1), img.nbytes, dtype=np.uint32)
+        data['modulesPresents'] = [True for i in range(self.modules)]
+        if self.corrected:
+            data['image.gain'] = np.zeros(self.data_shape, dtype=np.uint16)
+            data['passport'] = self.corr_passport()
+        data['pulseCount'] = self.pulses
+        # TODO: pulseId differ between AGIPD/LPD
+        data['pulseId'] = np.arange(self.pulses, dtype=np.uint16)
+        data['sources'] = sources
+        data['trainId'] = trainId
+
+        meta = self.gen_metadata(self.source, timestamp, trainId)
+        return {self.source: data}, meta
+
+
+class AGIPD(Detector):
+    pulses = 64
+    modules = 16
+    mod_y = 128
+    mod_x = 512
+    pixel_size = 0.2
+    distance = 2000
+    layout = np.array([
+        [12, 0],
+        [13, 1],
+        [14, 2],
+        [15, 3],
+        [8, 4],
+        [9, 5],
+        [10, 6],
+        [11, 7],
+    ])
+
+    const_fields = {
+        'checksum': b'\x00'*16,
+        'data': b'\x01'*416,
+        'dataId': 0,
+        'linkId': 18446744069414584335,
```

Comment:
Do these have a prefix, like `header.linkId`? Or is that only added when writing the files?

Reply:
Those are the keys given by the latest implementation (C++) of the calibration pipeline. We should maybe ask Steffen in the next meeting if we should reflect the name in files of at least define names that should not change between implementation.